src/util/newconfig/ConfigConstraints.cpp

```diff
@@ -0,0 +1,137 @@
+    }
+    if (p >= portMin && p <= portMax)
+        return std::nullopt;
+    return Error{"Port does not satisfy the constraint bounds"};
+}
+
+std::optional<Error>
+ChannelNameConstraint::checkConstraint(Constraint::ValueType const& channel) const
+{
+    if (!checkType(channel))
+        return Error{"Channel name must be a string"};
+
+    if (std::ranges::any_of(channels, [&channel](std::string_view name) {
+            return std::get<std::string>(channel) == name;
+        }))
+        return std::nullopt;
+    return Error{"Channel name must be one of General, WebServer, Backend, RPC, ETL, Subscriptions, Performance"};
+}
+
+std::optional<Error>
+LogLevelNameConstraint::checkConstraint(Constraint::ValueType const& logger) const
+{
+    if (!checkType(logger))
+        return Error{"log_level must be a string"};
+
+    if (std::ranges::any_of(logLevels, [&logger](std::string_view name) {
+            return std::get<std::string>(logger) == name;
+        }))
+        return std::nullopt;
+    return Error{"log_level must be one of trace, debug, info, warning, error, fatal, count"};
+}
+
+std::optional<Error>
+ValidIPConstraint::checkConstraint(Constraint::ValueType const& ip) const
+{
+    if (!checkType(ip))
+        return Error{"ip must be a string"};
+
+    static std::regex const ipv4Regex(
+        R"(^((25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$)"
+    );
+    if (std::regex_match(std::get<std::string>(ip), ipv4Regex))
+        return std::nullopt;
+    return Error{"ip is not a valid ip address"};
+}
+
+std::optional<Error>
+CassandraName::checkConstraint(ValueType const& name) const
+{
+    if (checkType(name) && std::get<std::string>(name) == "cassandra")
+        return std::nullopt;
+    return Error{"database.type must be string Cassandra"};
+}
+
+std::optional<Error>
+LoadConstraint::checkConstraint(ValueType const& name) const
+{
+    if (!checkType(name))
+        return Error{"cache.load must be a string"};
+
+    auto const load = std::get<std::string>(name);
+    if (load == "sync" || load == "async" || load == "none")
+        return std::nullopt;
+    return Error{"cache.load must be string sync, async, or none"};
+}
+
+std::optional<Error>
+LogTagStyle::checkConstraint(ValueType const& tagName) const
+{
+    if (!checkType(tagName))
+        return Error{"log_tag_style must be a string"};
+
+    if (std::ranges::any_of(logTags, [&tagName](std::string_view name) {
+            return std::get<std::string>(tagName) == name;
+        }))
+        return std::nullopt;
+    return Error{"log_tag_style must be string int, uint, null, none, or uuid"};
+}
+
+std::optional<Error>
+APIVersionConstraint::checkConstraint(ValueType const& apiVersion) const
+{
+    if (!checkType(apiVersion))
+        return Error{"api_version must be an integer"};
+
+    if (std::get<int64_t>(apiVersion) <= API_VERSION_MAX && std::get<int64_t>(apiVersion) >= API_VERSION_MIN)
```

Comment:
This one also share the same logic with constraint of port. They can abstract to one constraint, like "Between". It is more readable when we use it .  
like 
`ConfigValue{ConfigType::Integer}.defaultValue(1u).withConstraint(Between(1,3))`

Reply:
Thanks for the suggestion!! After some discussion, I think I'm going to stick with current impl. The Between(1,3) looks fine, but if it is for port, or uint16, it's going to look like withConstraint(Between(0, 65535)), which is hard to decipher what the constraint means